### DISCUSSION TITLE
Run middleware after web application check if the path exists

### DISCUSSION THREAD

**Discussion by tonygcs** [aio-libs/aiohttp#7326] [Q&A] [answered]
Hello,

I'm looking for a way to run my middleware after the application check if the path that the client requested exists or not. I don't want my middleware to run if the path does not exist and I have to return a `404`.

This is a server example:

```py
from aiohttp import web


@web.middleware
async def log_middleware(request, handler):
    print('Incoming request')
    response = await handler(request)
    print('Request handled')
    return response


async def hello_world(request):
    return web.Response(text="Hello World")


app = web.Application(middlewares=[log_middleware])
app.router.add_get("/valid", hello_world)
web.run_app(app)
```

So if the user request `http://localhost:8080/invalid` the server output is the following:

```sh
======== Running on http://0.0.0.0:8080 ========
(Press CTRL+C to quit)
Incoming request
```

Note that the server does not log the message `'Request handled'`.

The client gets the following response:

```sh
> curl http://localhost:8080/invalid
404: Not Found
```

My python version is `3.10.6` and the `aiohttp` version is `3.8.4`.

Do you have an idea about how we can imlpement this scenario?

Thanks in advance.

**Comment by Dreamsorcerer** [comment]
The handler is raising the 404 as an exception, that's why your second print doesn't happen. You can catch and reraise the exception, if you want to run something in that situation. This is demonstrated in the tutorial which customises the default error responses:
https://demos.aiohttp.org/en/latest/tutorial.html#middlewares

**Reply by tonygcs** [reply]
Thanks for the reply @Dreamsorcerer,

Sorry, maybe I'm missing something but In my case I don't want to run specific code when this exception happens.

Imagine that I have a middleware that runs a process that takes 10 seconds before it calls `response = await handler(request)`. It does not make sense to run it If the service is going to return a `404`. That's because I would like to run the middleware after the application checks if the path exists or not.

**Reply by Dreamsorcerer** [reply, marked as answer]
Looking at the code, try this:
`if isinstance(request.match_info.http_exception, web.HTTPNotFound):`

**Reply by tonygcs** [reply]
Thanks @Dreamsorcerer, it works as expected. I was able to build my own middleware that includes this _if_ statement and manage the invalid endpoint in the place that I want.
